### PR TITLE
8337280: Include jdk.jsobject module with JavaFX

### PR DESCRIPTION
**Pull request by kevinrushforth**
Add the `jdk.jsobject` module to JavaFX. This module is currently delivered by the JDK, but will be terminally-deprecated in  JDK 24 by [JDK-8311530](https://bugs.openjdk.org/browse/JDK-8311530), which is out for review at openjdk/jdk#20555. We therefore plan to start delivering it with JavaFX in JavaFX 24.

### Testing scenarios

Both JDK and JavaFX for some period of time, so we need to test various scenarios. I built and tested this PR branch with each of the following JDKs:

1. JDK 23 or earlier (I used the default boot JDK, which is currently JDK 22.0.2); this JDK includes `jdk.jsobject` as a normal, non-upgradable module
2. A locally-built JDK 24 from my [8311530-depr-jsobject](https://github.com/kevinrushforth/jdk/tree/8311530-depr-jsobject) branch, which has the fix for [JDK-8311530](https://bugs.openjdk.org/browse/JDK-8311530); this JDK includes `jdk.jsobject` as a terminally-deprecated, upgradable module
3. A locally-built JDK 24 from my [WIP-rm-jsobject](https://github.com/kevinrushforth/jdk/tree/WIP-rm-jsobject) branch, which removes the `jdk.jsobject` module, and is therefore a prototype of what will happen in JDK 26 (or later) when it is removed from the JDK

I ran a build and test of JavaFX from this PR branch using each of the above JDKs. All tests pass.

I then took the JavaFX artifacts built by JDK 1 (matching how we will deliver this in JavaFX 24), and ran the following additional tests, all of which match my expectations:

* java --module-path=.../jfx/build/sdk/lib without any additional flags (required for JDK 1; recommended for JDK 3) <br> RESULT: JDKs 1 and 2 use the `jdk.jsobject` module from the JDK; JDK 3 uses the `jdk.jsobject` module from JavaFX
* java --module-path=.../jfx/build/sdk/lib --upgrade-module-path=.../jfx/build/sdk/lib (recommended for JDK 2, OK for 3) <br> RESULT: JDK 1 will fail to launch; JDKs 2 and 3 use the `jdk.jsobject` module from JavaFX
* jlink with .../jfx/build/jmods ahead of $JAVA_HOME/jmods (recommended for JDKs 2 and 3) <br> RESULT: JDK 1 will fail to link; JDKs 2 and 3 use the `jdk.jsobject` module from JavaFX
* jlink with $JAVA_HOME/jmods ahead of .../jfx/build/jmods (required for JDK 1) <br> RESULT: JDKs 1 and 2 will use the `jdk.jsobject` module from the JDK; JDK 3 uses the `jdk.jsobject` module from JavaFX

I have not verified the maven artifacts, but they are being generated correctly.

---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change requires CSR request [JDK-8338249](https://bugs.openjdk.org/browse/JDK-8338249) to be approved
- [x] Change must be properly reviewed (2 reviews required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer), 1 [Author](https://openjdk.org/bylaws#author))

### Issues
 * [JDK-8337280](https://bugs.openjdk.org/browse/JDK-8337280): Include jdk.jsobject module with JavaFX (**Enhancement** - P4)
 * [JDK-8338249](https://bugs.openjdk.org/browse/JDK-8338249): Include jdk.jsobject module with JavaFX (**CSR**)

### Reviewers
 * [Ambarish Rapte](https://openjdk.org/census#arapte) (@arapte - **Reviewer**)
 * [Jay Bhaskar](https://openjdk.org/census#jbhaskar) (@jaybhaskar - Committer)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jfx.git pull/1529/head:pull/1529` \
`$ git checkout pull/1529`

Update a local copy of the PR: \
`$ git checkout pull/1529` \
`$ git pull https://git.openjdk.org/jfx.git pull/1529/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 1529`

View PR using the GUI difftool: \
`$ git pr show -t 1529`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jfx/pull/1529.diff">https://git.openjdk.org/jfx/pull/1529.diff</a>

</details>

### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jfx/pull/1529#issuecomment-2332645471)

### CODE REVIEW

File: modules/jdk.jsobject/src/main/java/netscape/javascript/JSObject.java [hunk 1]
```diff
@@ -0,0 +1,139 @@
...
+ * converted to Java data types. Certain data passed to the JavaScript
+ * engine is converted to JavaScript data types.
+ * </p>
+ * @since 1.5
+ */
+public abstract class JSObject {
+
+    static {
+        System.err.println("KCR: JavaFX version of jdk.jsobject loaded");
+    }
```
Comment (kevinrushforth) [inline modules/jdk.jsobject/src/main/java/netscape/javascript/JSObject.java:52]: This debug print statement is there to make it easy to see if the JavaFX version of this class is being loaded. I'll leave it during testing and review, but I'll remove it before the final review.

File: settings.gradle
Comment (AnirvanSarkar) [inline settings.gradle:1]: The update to copyright year has been reverted by commit [33355b4](https://github.com/openjdk/jfx/pull/1529/commits/33355b487e03f37ea369402c68e98554f81493cc).

**Comment by AnirvanSarkar** [comment]
Old version of classes `netscape.javascript.JSObject` and `netscape.javascript.JSException` seems to be delivered with JavaFX for Android [1].

Should these classes be removed?

[1] : https://github.com/openjdk/jfx/tree/jfx23/modules/javafx.web/src/android/java/netscape/javascript

File: settings.gradle
Comment (kevinrushforth) [reply settings.gradle:1]: I can't remember why I did that, but I'll redo the change to 2024. Thanks.

**Comment by kevinrushforth** [comment]
> Old version of classes `netscape.javascript.JSObject` and `netscape.javascript.JSException` seems to be delivered with JavaFX for Android [1].
> 
> Should these classes be removed?
> 
> [1] : https://github.com/openjdk/jfx/tree/jfx23/modules/javafx.web/src/android/java/netscape/javascript

I'll let @johanvos speak to that. If this is desired, then it's probably best done as a follow-up, since I can't test the Android (or iOS) implementation.

**Comment by kevinrushforth** [comment]
> I believe the scenario where jlink is used with an "old" JDK (e.g. 21) in combination with the most recent JavaFX release is very common. In that case, it seems to be required to prepend the module path with $JAVA_HOME/jmods, correct?

Yes.

> If that is the case, we need to announce and explain it very well, as developers will run into it and get confused.

Agreed. I've marked this RFE as needing a release note, and will be sure to clarify this.

**Comment by johanvos** [comment]
> > Old version of classes `netscape.javascript.JSObject` and `netscape.javascript.JSException` seems to be delivered with JavaFX for Android [1].
> > Should these classes be removed?
> > [1] : https://github.com/openjdk/jfx/tree/jfx23/modules/javafx.web/src/android/java/netscape/javascript
> 
> I'll let @johanvos speak to that. If this is desired, then it's probably best done as a follow-up, since I can't test the Android (or iOS) implementation.

Yes, those can be removed. There is more (ios/android) code that can be removed, or that even *should* be removed. I'll file a follow-up for this.

**Comment by johanvos** [comment]
> > > Old version of classes `netscape.javascript.JSObject` and `netscape.javascript.JSException` seems to be delivered with JavaFX for Android [1].
> > > Should these classes be removed?
> > > [1] : https://github.com/openjdk/jfx/tree/jfx23/modules/javafx.web/src/android/java/netscape/javascript
> > 
> > 
> > I'll let @johanvos speak to that. If this is desired, then it's probably best done as a follow-up, since I can't test the Android (or iOS) implementation.
> 
> Yes, those can be removed. There is more (ios/android) code that can be removed, or that even _should_ be removed. I'll file a follow-up for this.

Created https://bugs.openjdk.org/browse/JDK-8339705 for this

**Comment by kevinrushforth** [comment]
After merging in the latest master yesterday, I discovered that the recently-enabled javac lint warnings and `-Werror` when compiling the test classes will fail the build with "removal" warnings for those tests that use JSObject, when I run using my JDK 24 build with the patch from openjdk/jdk#20555. This is a good thing, since it means that those tests were not using the version of jdk.jsojbect from JavaFX. This highlights the needs to use `--upgradle-module-path` in our unit tests when running with JDK 24, so I pushed a commit that does that.

I'm done with my testing, and will switch to creating the CSRs.

Reviewers: @arapte @jaybhaskar

**Comment by kevinrushforth** [comment]
I've created the CSRs for both the this PR and openjdk/jdk#20555.

**Review by arapte — APPROVED** [review]

**Review by arapte — APPROVED** [review]

**Review by jaybhaskar — APPROVED** [review]
+looks good to me, The existing web test has setMember like tests.

**Comment by AnirvanSarkar** [comment]
I noticed that other JavaFX modules have `make/build.properties` file.
It seems this file is used when co-bundling JavaFX with the JDK, when JDK is configured to be built with `--with-import-modules` option.
Should we also add this file for `jdk.jsobject` module in JavaFX ?

**Comment by kevinrushforth** [comment]
Yes, I missed that. Thanks for catching it. I'll add a `make/build.properties` file, although the last time I checked, there were other problems in building a JavaFX bundle that could be used with `--with-import-modules`, so it may not be sufficient.

**Review by arapte — APPROVED** [review]

**Review by jaybhaskar — APPROVED** [review]
Looks good to me.

**Pull request closed without merging**